src/mame/yamaha/ymdx9.cpp

```diff
@@ -0,0 +1,449 @@
+	 *  Bit 1: "".
+	 *  Bit 2: "".
+	 *  Bit 3: "".
+	 *  Bit 4: ADC EOC Input Line.
+	 *  Bit 5: Cassette Interface Remote Port.
+	 *  Bit 6: Cassette Interface Tape Output.
+	 *  Bit 7: Cassette Interface Tape Input.
+	 * @param offset The offset into the memory mapped region being written.
+	 * @param data The data being written.
+	 */
+	void p1_w(offs_t offset, uint8_t data);
+
+	/**
+	 * @brief Handle a read from the synth's IO Port 1.
+	 * @param offset The offset into the memory mapped region being read.
+	 * @return uint8_t The value read from the port.
+	 */
+	uint8_t p1_r(offs_t offset);
+
+	void unmapped_w(offs_t address, uint8_t data);
+
+	uint8_t unmapped_r(offs_t offset);
+};
+
+
+/**
+* yamaha_dx9_state::machine_start
+*/
+void yamaha_dx9_state::machine_start()
+{
+	m_leds.resolve();
+	m_rx_data = ASSERT_LINE;
+}
+
+
+/**
+ * yamaha_dx9_state::lcd_pixel_update
+ */
+HD44780_PIXEL_UPDATE(yamaha_dx9_state::lcd_pixel_update)
+{
+	if (x < 5 && y < 8 && line < 2 && pos < 16)
+			bitmap.pix(line * 10 + y + 1 + ((y == 7) ? 1 : 0), pos * 6 + x + 1) = state ? 1 : 2;
+}
+
+
+/**
+ * yamaha_dx9_state::palette_init
+ */
+void yamaha_dx9_state::palette_init(palette_device &palette)
+{
+	palette.set_pen_color(0, rgb_t(0x87, 0xAD, 0x34)); // background
+	palette.set_pen_color(1, rgb_t(0x0, 0x0, 0x0)); // lcd pixel on
+	palette.set_pen_color(2, rgb_t(0x7D, 0x9F, 0x32)); // lcd pixel off
+}
+
+
+/**
+ * yamaha_dx9_state::mem_map
+ */
+void yamaha_dx9_state::mem_map(address_map &map)
+{
+	// Internal CPU registers.
+	map(0x0000, 0x001f).m(m_maincpu, FUNC(hd6303r_cpu_device::m6801_io));
+
+	map(0x20, 0x20).r(FUNC(yamaha_dx9_state::key_switch_scan_driver_r));
+
+	map(0x22, 0x22).r("adc", FUNC(m58990_device::data_r));
+	map(0x24, 0x24).w("adc", FUNC(m58990_device::address_data_start_w));
+
+	// YM21280 OPS.
+	map(0x26, 0x27).w(FUNC(yamaha_dx9_state::ops_w));
+	// HD44780 LCD Controller.
+	map(0x0028, 0x0029).rw("lcdc", FUNC(hd44780_device::read), FUNC(hd44780_device::write));
+	// LED.
+	map(0x2B, 0x2C).w(FUNC(yamaha_dx9_state::led_w));
+
+	// Internal RAM.
+	map(0x0040, 0x00ff).ram();
+
+	// External RAM.
+	// 2 * 2kb RAM1 IC19 M5M118P.
+	map(0x0800, 0x0fff).ram().share("ram1");
+	map(0x1000, 0x1800).ram().share("ram2");
+
+	// YM21290 EGS
+	map(0x1800, 0x18F3).w(FUNC(yamaha_dx9_state::egs_w));
+
+	// ROM.
+	map(0xc000, 0xffff).rom().region("program", 0);
+}
+
+
+/**
+ * yamaha_dx9_state::dx9
+ */
+void yamaha_dx9_state::dx9(machine_config &config)
+{
+	// Initialise the HD63B03RP CPU.
+	// This oscillator frequency comes from the service manual.
+	HD6303R(config, m_maincpu, 9.4265_MHz_XTAL / 2);
+	m_maincpu->set_addrmap(AS_PROGRAM, &yamaha_dx9_state::mem_map);
+
+	// Unlike the DX7 only IO port 1 is used.
+	// The direction flags of other ports are set, however they are never read, or written.
+	m_maincpu->in_p1_cb().set(FUNC(yamaha_dx9_state::p1_r));
+	m_maincpu->in_p2_cb().set(FUNC(yamaha_dx9_state::unmapped_r));
```

Comment:
If the port isn't used at all don't map it. You can use `set_constant` if it requires a specific value.

Reply:
I've removed the unnecessary mapping. I'm new to MAME development, and based this work off the DX7 skeleton driver, which is in a stale PR. I got these functions from that PR. Thinking of Chesterton's fence, I figured maybe these were required for some reason.